tests/math/test_functions.py

```diff
@@ -1392,3 +1441,178 @@ def test_torch_coercion(self):
         res = qml.math.coerce(tensors, like="torch")
         dtypes = [r.dtype for r in res]
         assert all(d is torch.complex64 for d in dtypes)
+
+
+class TestUnwrap:
+    """Test tensor unwrapping"""
+
+    def test_tensorflow_unwrapping(self):
+        """Test that a sequence of TensorFlow values is properly unwrapped"""
+        values = [
+            onp.array([0.1, 0.2]),
+            tf.Variable(0.1, dtype=tf.float64),
+            tf.constant([0.5, 0.2]),
+        ]
+        res = qml.math.unwrap(values)
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(res, expected))
+
+    def test_torch_unwrapping(self):
+        """Test that a sequence of Torch values is properly unwrapped"""
+        values = [
+            onp.array([0.1, 0.2]),
+            torch.tensor(0.1, dtype=torch.float64),
+            torch.tensor([0.5, 0.2]),
+        ]
+        res = qml.math.unwrap(values)
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(res, expected))
+
+    def test_autograd_unwrapping_forward(self):
+        """Test that a sequence of Autograd values is properly unwrapped
+        during the forward pass"""
+        unwrapped_params = None
+
+        def cost_fn(params):
+            nonlocal unwrapped_params
+            unwrapped_params = qml.math.unwrap(params)
+            return np.sum(np.sin(params[0] * params[2])) + params[1]
+
+        values = [onp.array([0.1, 0.2]), np.tensor(0.1, dtype=np.float64), np.tensor([0.5, 0.2])]
+        cost_fn(values)
+
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(unwrapped_params, expected))
+        assert all(not isinstance(a, np.tensor) for a in unwrapped_params)
+
+    def test_autograd_unwrapping_backward(self):

```

Comment:
just curious, why does autograd need backward and forward testing, but not the other interfaces?

Reply:
Because Autograd behaves differently during the backward and forward passes :(

- On the forward pass, it simply passes `np.tensor` objects. They are unwrapped by calling `.numpy()`.

- On the backward pass, there is an implicit forward pass where it instead passes `ArrayBox` objects. They are unwrapped by calling `._value`.

There is logic in `qml.math` to account for both cases, and I want to ensure it works correctly.

TF and Torch are different:

- they do not have a functional user-interface, so the forward pass and the backward pass always occur separately
- As a result, the forward pass is always explicit, and always just receives a tensor.